client/js/models/archive.js

```diff
@@ -192,53 +184,34 @@ module.exports = {
           if (file && file.progressListener && file.progressHandler) {
             file.progressListener.removeListener('progress', file.progressHandler)
           }
-          if (file.fullPath === '/dat.json') send('archive:updateMetadata', {}, noop)
+          if (file.fullPath === '/dat.json' || file.fullPath === '/datapackage.json') send('archive:updateMetadata', {}, noop)
           send('archive:updateImportQueue', {onFileWriteComplete: true}, noop)
         }
       })
     },
     load: function (key, state, send, done) {
-      var archive, sw, timer
-      if (state.instance && state.instance.drive) {
-        if (state.instance.key.toString('hex') === key) {
-          archive = state.instance
-          sw = state.swarm
-        } else {
-          archive = null
-        }
-      }
-      if (!archive) {
-        archive = drive.createArchive(key, {live: true, sparse: true})
-        sw = swarm(archive, {signalhub: state.signalhubs})
-        send('archive:update', {instance: archive, swarm: sw, key}, done)
+      if (state.instance) {
+        if (state.instance.key.toString('hex') === key) return done()
       }
-      sw.on('connection', function (conn) {
-        send('archive:updatePeers', noop)
-        conn.on('close', function () {
-          send('archive:updatePeers', noop)
-        })
-      })
-      archive.on('upload', function (data) {
-        send('archive:updateUploaded', data.length, noop)
-        if (timer) window.clearTimeout(timer)
-        timer = setTimeout(() => send('archive:update', {uploadSpeed: 0, downloadSpeed: 0}, noop), 3000)
-      })
-      archive.on('download', function (data) {
-        send('archive:updateDownloaded', data.length, noop)
-        send('archive:updateMetadata', {}, noop)
-        if (timer) window.clearTimeout(timer)
-        timer = setTimeout(() => send('archive:update', {uploadSpeed: 0, downloadSpeed: 0}, noop), 3000)
-      })
-      archive.open(function () {
-        if (archive.content) {
-          archive.content.get(0, function (data) {
-            send('archive:update', {size: archive.content.bytes}, noop)
-            // XXX: Hack to fetch a small bit of data so size properly updates
-          })
-        }
-        send('archive:updateMetadata', {}, noop)
+
+      var dat = Dat(drive, key, state.signalhubs, send)
+      key = dat.archive.key.toString('hex')
+      const location = '/' + key
+      send('location:setLocation', { location }, noop)
+      window.history.pushState({}, null, location)
+      var stream = dat.archive.list({live: true})
+      stream.on('data', function (entry) {
+        var entries = state.entries
+        entries.push(entry)
+        send('archive:update', {entries}, noop)
       })
-      return done()
```

Comment:
it's in the callback to the final `send('archive:update', newState, done)`, and since `update` is synchronous, done is called when it's done.